testsrc/org/mozilla/javascript/tests/JavaAdapterInvokeTest.java

```diff
@@ -0,0 +1,111 @@
+        cx.setOptimizationLevel(-1);
+        topScope = cx.initStandardObjects();
+    }
+
+    @After
+    public void exitContext() {
+        Context.exit();
+    }
+
+    public interface AdapterInterface {
+        int m1(int i);
+
+        int m2();
+    }
+
+    public static class AdapterClass {
+        private AdapterInterface adapter;
+
+        public AdapterClass(AdapterInterface adapter) {
+            this.adapter = adapter;
+        }
+
+        public int doIt(int i) {
+            return this.adapter.m1(i) + this.adapter.m2();
+        }
+    }
+
+    @Test
+    public void testInvoke() throws NoSuchMethodException {
+        String testCode =
+                "'use strict'\n"
+                        + "var impl = {"
+                        + "  m1: function(i) { return i + 1 },\n"
+                        + "  m2: function() { return 7 }\n"
+                        + "}\n"
+                        + "adapter = new Packages."
+                        + AdapterClass.class.getName()
+                        + "(impl)\n"
+                        + "adapter.doIt(42)";
+
+        Number result = (Number) cx.evaluateString(topScope, testCode, "", 1, null);
+        Assert.assertEquals(50, result.intValue());
+    }
+
+    @Test
+    public void testInvokeWithPrototype() throws NoSuchMethodException {
+        String testCode =
+                "'use strict'\n"
+                        + "function Obj() {}\n"
+                        + "Obj.prototype.m1 = function(i) { return i + 1 }\n"
+                        + "Obj.prototype.m2 = function() { return 7 }\n"
+                        + "var impl = new Obj()\n"
+                        + "adapter = new Packages."
+                        + AdapterClass.class.getName()
+                        + "(impl)\n"
+                        + "adapter.doIt(42)";
+
+        Number result = (Number) cx.evaluateString(topScope, testCode, "", 1, null);
+        Assert.assertEquals(50, result.intValue());
+    }
+
+    @Test
+    public void testInvokeWithPrototypeAndCtor() throws NoSuchMethodException {
+        String testCode =
+                "'use strict'\n"
+                        + "function Obj() { this.myObj = {one: 1} }\n"
+                        + "Obj.prototype.m1 = function(i) { return i + this.myObj.one }\n"
+                        + "Obj.prototype.m2 = function() { return 7 }\n"
+                        + "var impl = new Obj()\n"
+                        + "adapter = new Packages."
+                        + AdapterClass.class.getName()
+                        + "(impl)\n"
+                        + "adapter.doIt(42)";
+
+        Number result = (Number) cx.evaluateString(topScope, testCode, "", 1, null);
+        Assert.assertEquals(50, result.intValue());
+    }
+
+    @Test
+    public void testInvokeJsOnly() throws NoSuchMethodException {
+        String testCode =
+                "'use strict'\n"
+                        + "function Obj() { this.myObj = {one: 1} }\n"
```

Comment:
What about a testcase where .myObj is added on the prototype, instead of the `this` inside the constructor?

Reply:
Didn't add a separate testcase, but added a second property "seven" on the prototype.
(So testcase will check both code paths at once - but FYI: the second path will also fail without the fix)